Add dealiasing support to `convolve_fft`

WilliamJamieson opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

Dealiasing pseudospectral convolutions such as `convolve_fft` can be important for the numerical stability of some numerical processes taking advantage of them. The simplest approach to dealiasing in this context is to zero pad the vectors prior to convolution and then to remove the data in the padding; in this case we implement this using the 2/3s scheme proposed by [Orszag](https://doi.org/10.1175/1520-0469(1971)028%3C1074:OTEOAI%3E2.0.CO;2). A good overview of dealiasing can be found in a presentation by [Bowman](http://helper.ipam.ucla.edu/publications/mtws1/mtws1_12187.pdf).

Dealiasing is implemented as an optional (default is off) argument to `convolve_fft`.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->
>

WilliamJamieson commented:
Tagging @perrygreenfield and @nden.

pllim commented on astropy/convolution/convolve.py:
```diff
@@ -384,7 +384,7 @@ def convolve_fft(array, kernel, boundary='fill', fill_value=0.,
                  normalization_zero_tol=1e-8,
                  preserve_nan=False, mask=None, crop=True, return_fft=False,
-                 fft_pad=None, psf_pad=None, min_wt=0.0, allow_huge=False,
+                 fft_pad=None, psf_pad=None, dealias=None, min_wt=0.0, allow_huge=False,
```
Can we move the new keyword to the end of the signature, in case some people are calling this by passing keyword as positional?

WilliamJamieson replied:
Done in 9fd8242

pllim commented:
I know this is a draft, so I assume tests and change log will come later? 😸

WilliamJamieson commented:
> I know this is a draft, so I assume tests and change log will come later? 😸

Yes!

dhomeier commented on astropy/convolution/convolve.py:
```diff
@@ -653,13 +663,19 @@ def convolve_fft(array, kernel, boundary='fill', fill_value=0.,
         if fft_pad:
             raise ValueError("With boundary='wrap', fft_pad cannot be enabled.")
         fft_pad = False
+        if fft_pad:
```
Did you intend to check for `dealias` here?

WilliamJamieson replied:
Fixed in cfaa006.

dhomeier commented on astropy/convolution/convolve.py:
```diff
@@ -670,20 +686,37 @@ def convolve_fft(array, kernel, boundary='fill', fill_value=0.,
+            if dealias:
+                bsize += np.ceil(bsize / 2)
+
+            fsize = 2 ** np.ceil(np.log2(np.max(bsize)))
```
```suggestion
            fsize = 2 ** np.ceil(np.log2(bsize))
```
`bsize` should already be a single scalar here as on L705, right?

keflavich commented:
Overall, seems like a useful feature and a good thing to have available.  However:

This PR will have major conflict with https://github.com/astropy/astropy/pull/11533 (they edit the same parts of the code), which takes precedence.

There are also some problems with this implementation that I'd like to see addressed:

- Zero-padding by 1/3 is recommended in the docstring, but by 1/2 is implemented, as far as I can tell - unless the 1/3 is measured on the final size?
- in most cases, the zero-padding will be excessive, since both PSF and FFT padding naturally perform zero-padding.  I would like to see a check that, _if_ the padding is inadequate, _then_ additional padding will be added.
- Additional padding has major performance implications (see https://github.com/astropy/astropy/issues/11242#issuecomment-818017906 and https://github.com/astropy/astropy/issues/11242#issuecomment-760169792).  These need to be mentioned in the docstring.
- It would be really helpful to add user-facing documentation explaining when this additional de-aliasing is necessary.  Was there a specific use case that brought you to write this PR?  If so, describe it as a clear example!  The Bowman paper points out the need in the specific context of modeling turbulence, but I strongly suspect the astropy routine is a terrible choice for doing FFT convolutions in the context of turbulence simulations.

WilliamJamieson commented:
> Overall, seems like a useful feature and a good thing to have available. However:
> 
> This PR will have major conflict with #11533 (they edit the same parts of the code), which takes precedence.
> 

Not a problem, I can handle fixing the conflicts once #11533 gets merged in. All the useful tests that I can construct rely on stuff implemented in #11512, so I was planning on having to fix conflicts prior to this being merged anyways.

> There are also some problems with this implementation that I'd like to see addressed:
> 
> * Zero-padding by 1/3 is recommended in the docstring, but by 1/2 is implemented, as far as I can tell - unless the 1/3 is measured on the final size?

The 1/3 is indeed measured on the padded size. Essentially, to dealias one needs to throw out at least 1/3 of what is returned by the pseudospectral convolution; hence, one extends by 1/2 so that the input to the convolution is 3/2 meaning when we remove 1/3 of the result we are left with the original length input.

> * in most cases, the zero-padding will be excessive, since both PSF and FFT padding naturally perform zero-padding.  I would like to see a check that, _if_ the padding is inadequate, _then_ additional padding will be added.

When I originally implemented this, we were always padding to be a power of 2 length. Assuming that the lengths of inputs are random, this means that about 1/2 the time for the FFT padding alone will pad more than the dealiasing padding. The way I implemented dealiasing padding was to pad the input value prior to passing it to the FFT padding, which means the 1/2 the time the arrays won't get extended any further than we were already seeing. Similar arguments can be made for the PSF padding. What this means is that many times `convolve_fft` was dealiasing by accident. Turning on dealiasing will just ensure that all inputs would be dealiased.

> * Additional padding has major performance implications (see [#11242 (comment)](https://github.com/astropy/astropy/issues/11242#issuecomment-818017906) and [#11242 (comment)](https://github.com/astropy/astropy/issues/11242#issuecomment-760169792)).  These need to be mentioned in the docstring.

Agreed, additional padding can have major performance implications. However, as noted above PSF and FFT padding often takes care of the additional dealiasing, so depending on data sizes we may see no performance changes. There are other methods of dealiasing including phase-shifting and implicit dealiasing. I implemented zero-padding because we already had the infrastructure in place to handle zero-padding so it was a simple thing to fit in. Phase-shifting is a little harder to implement, but doesn't gain you that much performance. Lots can be gained by Bowman's implicit dealiasing approach, but that requires a switch of FFT algorithms to FFTW++. If one is so concerned they can use the python wrapper for FFTW++ and pass it in a the FFT being used.

> * It would be really helpful to add user-facing documentation explaining when this additional de-aliasing is necessary.  Was there a specific use case that brought you to write this PR?  If so, describe it as a clear example!  The Bowman paper points out the need in the specific context of modeling turbulence, but I strongly suspect the astropy routine is a terrible choice for doing FFT convolutions in the context of turbulence simulations.

Honestly, my background is in computational fluid dynamics and so the only examples I know are all related to what Bowman describes. All of them amount to performing a numerical process where one continually feeds the results back through the convolution, such as when time-integrating to solve a PDE. Basically, the aliasing error amounts to energy being erroneously transferred  from the high frequencies to the low frequencies in Fourier space, meaning the convolution is not properly damping the solution over time. 

The reason that I even thought to include it here was that I was asked to look at issues related to `convolve_models` and noticed that there was not any attempt at dealiasing in `convolve_fft`, which was standard practice whenever I was doing CFD work. As `convolve_fft` already had the infrastructure to properly handle zero-padding I implemented it because it was relatively simple, and thought it couldn't hurt to add it in for completeness.

dhomeier commented:
> > This PR will have major conflict with #11533 (they edit the same parts of the code), which takes precedence.
> 
> Not a problem, I can handle fixing the conflicts once #11533 gets merged in. All the useful tests that I can construct rely on stuff implemented in #11512, so I was planning on having to fix conflicts prior to this being merged anyways.
> 
Yes; there will be merge conflicts, but once rebased (#11533 has just been merged) resolving them should be straightforward enough. I have already restructured the `psf_pad` and `fft_pad` sections more along the lines of how this PR implements them, and you should then be able to simply apply your padding to the sizes returned from the `psf_pad` block and then pass them on to `fft_pad`.

> > There are also some problems with this implementation that I'd like to see addressed:
> > 
> > * Zero-padding by 1/3 is recommended in the docstring, but by 1/2 is implemented, as far as I can tell - unless the 1/3 is measured on the final size?
> 
> The 1/3 is indeed measured on the padded size. Essentially, to dealias one needs to throw out at least 1/3 of what is returned by the pseudospectral convolution; hence, one extends by 1/2 so that the input to the convolution is 3/2 meaning when we remove 1/3 of the result we are left with the original length input.
> 
> > * in most cases, the zero-padding will be excessive, since both PSF and FFT padding naturally perform zero-padding.  I would like to see a check that, _if_ the padding is inadequate, _then_ additional padding will be added.
> 
> When I originally implemented this, we were always padding to be a power of 2 length. Assuming that the lengths of inputs are random, this means that about 1/2 the time for the FFT padding alone will pad more than the dealiasing padding. The way I implemented dealiasing padding was to pad the input value prior to passing it to the FFT padding, which means the 1/2 the time the arrays won't get extended any further than we were already seeing. Similar arguments can be made for the PSF padding. What this means is that many times `convolve_fft` was dealiasing by accident. Turning on dealiasing will just ensure that all inputs would be dealiased.
> 
Yes, this approach is much preferable, applying just the amount that is needed rather than relying on some chance numbers from the `fft_pad` scheme. Plus, with the new `fft_pad` the sizes are typically only increased by some 1-5 % for the `real=True` case, and <1 % for `real=False`, so dealiasing will definitely need extra padding.
For `psf_pad` all depends of course on kernel size, which in many cases also will be small compared to the input data. But it would be good to clarify if the dealiasing padding indeed needs to be done on top of that, as in the current implementation, or could just be the larger of the two.

> > * Additional padding has major performance implications (see [#11242 (comment)](https://github.com/astropy/astropy/issues/11242#issuecomment-818017906) and [#11242 (comment)](https://github.com/astropy/astropy/issues/11242#issuecomment-760169792)).  These need to be mentioned in the docstring.
> 
> Agreed, additional padding can have major performance implications. However, as noted above PSF and FFT padding often takes care of the additional dealiasing, so depending on data sizes we may see no performance changes. There are other

This will definitely no longer be the case; as noted above, dealiasing would now usually become the biggest factor in increasing data sizes, but certainly worth it. But this should be pointed out appropriately in the docs, in particular regarding the increase in memory usage, which for multi-dimensioned inputs could become fairly large.

> methods of dealiasing including phase-shifting and implicit dealiasing. I implemented zero-padding because we already had the infrastructure in place to handle zero-padding so it was a simple thing to fit in. Phase-shifting is a little harder to implement, but doesn't gain you that much performance. Lots can be gained by Bowman's implicit dealiasing approach, but that requires a switch of FFT algorithms to FFTW++. If one is so concerned they can use the python wrapper for FFTW++ and pass it in a the FFT being used.
> 

This would also be great to have a brief description in the docs, like the one for using `PyFFTW3`.

WilliamJamieson commented:
I just rebased and fixed merge conflicts so that #11533's changes can be incorporated.

dhomeier commented:
> > * Zero-padding by 1/3 is recommended in the docstring, but by 1/2 is implemented, as far as I can tell - unless the 1/3 is measured on the final size?
> 
> The 1/3 is indeed measured on the padded size. Essentially, to dealias one needs to throw out at least 1/3 of what is returned by the pseudospectral convolution; hence, one extends by 1/2 so that the input to the convolution is 3/2 meaning when we remove 1/3 of the result we are left with the original length input.
> 

But that sounds really reasoning backwards – I think most people would understand "pad the image by 1/3" as "add ⅓ of the original image size". I would also state explicitly here that this will increase the size of the N-dimensional input by about `1.5**N`.

WilliamJamieson commented:
> > > * Zero-padding by 1/3 is recommended in the docstring, but by 1/2 is implemented, as far as I can tell - unless the 1/3 is measured on the final size?
> > 
> > 
> > The 1/3 is indeed measured on the padded size. Essentially, to dealias one needs to throw out at least 1/3 of what is returned by the pseudospectral convolution; hence, one extends by 1/2 so that the input to the convolution is 3/2 meaning when we remove 1/3 of the result we are left with the original length input.
> 
> But that sounds really reasoning backwards – I think most people would understand "pad the image by 1/3" as "add ⅓ of the original image size". I would also state explicitly here that this will increase the size of the N-dimensional input by about `1.5**N`.

The reasoning seems backwards because the theorem concerning aliasing by the DFT in pseudospectral convolutions states that 1/3 of the data resulting from such a convolution will have some sort of aliasing error. This theorem makes no mention of padding or other pre and post post processing of the results. So the dealiasing rule gets referred to in terms of the 1/3, even though the solution is to pad the input data by 1/2 so that one can throw out "1/3" of the convolution results.

I may have explained what is going on poorly, and have taken steps to correct this in my latest changes to the documentation in: 12bab2d.

I am also working with Bowman, et. al. to create a relevant set of examples and use cases for dealiasing in astropy; however, it maybe sometime before we have something to add in.

dhomeier commented on astropy/convolution/convolve.py:
```diff
@@ -546,6 +546,12 @@ def convolve_fft(array, kernel, boundary='fill', fill_value=0.,
         Which complex dtype to use.  `numpy` has a range of options, from 64 to
         256.
+    dealias: bool, optional
+        Default off. Zero-pad image to allow enable explicit dealiasing
```
```suggestion
        Default off. Zero-pad image to enable explicit dealiasing
```
"allow" xor "enable"

dhomeier commented on astropy/convolution/convolve.py:
```diff
@@ -546,6 +546,12 @@ def convolve_fft(array, kernel, boundary='fill', fill_value=0.,
+    dealias: bool, optional
+        Default off. Zero-pad image to allow enable explicit dealiasing
+        of convolution. With ``boundary='wrap'``, this will be disabled.
+        Note that for an input of size N (entries) this will increase
+        the size of the computed convolution by at least 1.5*N. This
+        may result in significantly more memory usage.
```
```suggestion
        Note that for an input of nd dimensions this will increase
        the size of the temporary arrays by at least ``1.5**nd``.
        This may result in significantly more memory usage.
```
Slight misunderstanding. I have switched to 'nd' which is the terminology at the top of the docstring.
The last line could perhaps better be added to the Notes following the one on `psf_pad` impact.

dhomeier commented:
> The reasoning seems backwards because the theorem concerning aliasing by the DFT in pseudospectral convolutions states that 1/3 of the data resulting from such a convolution will have some sort of aliasing error. This theorem makes no mention of padding or other pre and post post processing of the results. So the dealiasing rule gets referred to in terms of the 1/3, even though the solution is to pad the input data by 1/2 so that one can throw out "1/3" of the convolution results.
> 
> I may have explained what is going on poorly, and have taken steps to correct this in my latest changes to the documentation in: [12bab2d](https://github.com/astropy/astropy/commit/12bab2d7106be15fd275c5058fa51e3bb4056a0c).
> 
> I am also working with Bowman, et. al. to create a relevant set of examples and use cases for dealiasing in astropy; however, it maybe sometime before we have something to add in.

OK, if the '⅓'-rule is a familiar concept that might in fact make sense, but I think this is best explained in this more extensive discussion under `Notes`. That might perhaps go better into the actual docs, perhaps in `using.rst`, but I'll let the maintainers chime in on this.

WilliamJamieson commented:
> I also saw in the Roberts & Bowman paper that the ⅓ | 50 % extension applies for centred convolutions, but then according to the docs the kernel here is always _assumed_ to be centred, so I guess this is OK?

This is my understanding as well.

eteq commented:
Since this hasn't made the feature freeze cutoff I'm re-milestoning to 5.0

keflavich reviewed (approved):
This seems fine now; I had lost track of it for a while but I think we have resolved the main issues, and there is enough discussion about when you might want to dealias in the docstring.

nden commented:
@larrybradley , @dhomeier Do you plan to review this further?

larrybradley commented on astropy/convolution/convolve.py:
```diff
@@ -432,7 +432,7 @@ def convolve_fft(array, kernel, boundary='fill', fill_value=0.,
                  fft_pad=None, psf_pad=None, min_wt=0.0, allow_huge=False,
                  fftn=np.fft.fftn, ifftn=np.fft.ifftn,
-                 complex_dtype=complex):
+                 complex_dtype=complex, dealias=None):
```
```suggestion
                 complex_dtype=complex, dealias=False):
```

larrybradley commented on astropy/convolution/convolve.py:
```diff
@@ -695,6 +719,8 @@ def convolve_fft(array, kernel, boundary='fill', fill_value=0.,
             psf_pad = True
         if fft_pad is None:
             fft_pad = True
+        if dealias is None:
+            dealias = False
```
```suggestion
```

larrybradley commented on astropy/convolution/convolve.py:
```diff
@@ -705,13 +731,18 @@ def convolve_fft(array, kernel, boundary='fill', fill_value=0.,
         if fft_pad is None:
             # default is 'True' according to the docstring
             fft_pad = True
+        if dealias is None:
+            # default is 'False' according to the docstring
+            dealias = False
```
```suggestion
```

larrybradley reviewed (changes requested):
Since `dealias` is a boolean, I think the default should be `False` instead of `None`.

nstarman commented:
@larrybradley, is this ready to be merged?

larrybradley commented:
@dhomeier Would you like to re-review this before merging?
